<!DOCTYPE html>
<html lang="en">

<head></head>

<body>

  <table class="main-table">
    <tr>
      <th>Project</th>
      <th>Cell line</th>
      <th>Description</th>
    </tr>

    <tr>
      <td><a href="?HEPG2">Liver Carcinogenicity</a></td>
      <td>HEPG2</td>
      <td>This experiment uses 330 selected chemicals for in-vivo liver carcinogenicity testing, including 128 liver carcinogens, 168 non-carcinogens, and 34 miscellaneous chemicals (e.g., nuclear receptor ligands). Chemical carcinogenicity and genotoxicity annotations are mainly based on the <em>Carcinogenicity Potency Database</em> (<a href="https://www.library.ucdavis.edu/database/carcinogenic-potency-database-cpdb/">CPDB</a>), which is the result of tissue-specific long-term carcinogenicity tests in rodents. In this study, HepG2 (liver) cells were exposed to each individual chemical for 24 hours and their gene expression profiled on the L1000 platform. Each chemical was assayed at 6 doses (2-fold dilutions starting from the highest concentration of 40uM or 20uM) with triplicate profiles generated for each dose...</td>
    </tr>

    <tr>
      <td><a href="?MCF10A">Breast Carcinogenicity</a></td>
      <td>MCF10A</td>
      <td>This experiment uses 345 selected chemicals for breast carcinogenicity testing, including 120 breast carcinogens, 114 non-carcinogens, and 68 miscellaneous chemicals (e.g., nuclear receptor ligands, <a href="https://www.bu.edu/sph/about/departments/environmental-health/research/research-groups-and-centers/superfund-research-program-at-boston-university/">BUSRP</a> chemicals, lung carcinogens). Chemical carcinogenicity and genotoxicity annotations are based on the <em>Carcinogenicity Potency Database</em> (<a href="https://www.library.ucdavis.edu/database/carcinogenic-potency-database-cpdb/">CPDB</a>), which is the result of tissue-specific long-term carcinogenicity tests in rodents, or breast carcinogens from [<a href="http://dx.doi.org/10.1002/cncr.22656">Rudel et al., 2007</a>]. In this study, MCF10A (breast epithelial) cells were exposed to each individual chemical for 24 hours and their gene expression  profiled on the L1000 platform. Each chemical was assayed at 3 doses (3-fold dilutions starting from the highest concentration of 100uM, with the exception of selected BUSRP chemicals)...</td>
    </tr>

    <tr>
      <td><a href="?ADIPO">Adipogens</a></td>
      <td>3T3-L1</td>
      <td>This experiment uses 3T3-L1 cells differentiated in the presence of 77 chemicals (negative controls, nuclear receptor ligands known to influence adipocyte biology, suspected environmental PPARγ ligands) to group chemicals based on their effects on PPARγ’s transcriptome and downstream metabolic functions. The profiled chemicals include 38 known PPARγ ligands or modifying compounds, 22 negative controls, and 17 suspected PPARγ ligands/modifiers. 3' digital gene expression profiling was performed on mouse NIH 3T3-L1 pre-adipocytes exposed over a 10-day period during differentiation...</td>
    </tr>
  </table>

</body>
</html>
