---
output: html_document
---

## API EXPLORER

The API Explorer provides documentation on how users can retrieve data from Xposome application and create their next project. All successful responses are returned in JSON. Only queries that respond with a 200 response code is considered successful. 

Here are standard HTTP codes you will find in the “Status” element of the response body. 

<table class="status-table">
  <tr>
    <td><strong>Status Code</strong></td>
    <td><strong>Description</strong></td>
  </tr>
  <tr>
    <td>200 OK</td>
    <td>Standard HTTP successful response</td>
  </tr>
  <tr>
    <td>404 Bad Request or Source Not Found.</td>
    <td>Standard HTTP invalid request response</td>
  </tr>
  <tr>
    <td>500 Internal Server Error</td>
    <td>There was an unexpected error on our server. If you see this please contact <a href="mailto:montilab@bu.edu">montilab@bu.edu</a></td>
  </tr>
</table>

<br>

There are four data sources that are returned from the API.

<table class="source-table">
  <tr>
    <td><strong>Data</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>Portals</td>
    <td>a list of portals in the GeneHive Database</td>
    <td>list</td>
  </tr>
  <tr>
    <td>Chemicals</td>
    <td>a list of chemicals available in a specific portal</td>
    <td>list</td>
  </tr>
  <tr>
    <td>Statistics</td>
    <td>a summary statistic for a specific chemical based on gene expression, gene set enrichment analysis and gene connectivity</a></td>
    <td>data frame</td>
  </tr>
  <tr>
    <td>Datasets</td>
    <td>provide the data source for the Xposome portals such as profile and chemical annotation, expression set, gene set enrichent, etc.</a></td>
    <td>data frame</td>
  </tr>
</table>


Below are a list of R packages and GET requests in R to retrieve data from Xposome portals. However, one can use Python or other programs to implement the REST API calls.

```{r api-packages}

# R packages required for API calls
library(jsonlite)
library(httr)

```

### Portals

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/portals</td>
    <td>Return a list of portals available in the GeneHive database</td>
  </tr>
</table>
  
<br>

#### Implementation

https://montilab.bu.edu/Xposome-API/portals?orderby=asc

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>orderby</td>
    <td>asc (default) or desc</td>
    <td>Sort options</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-portals}

# url for local testing
url <- "https://montilab.bu.edu/Xposome-API/portals?orderby=asc"

# Send GET Request to API
res <- GET(url = url, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

### If GET request is successful, return the results
if(test_request == "pass"){
  
  portals <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(portals))
  
}

```

### Chemicals

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/chemicals</td>
    <td>Return a list of chemicals available in a specific portal</td>
  </tr>
</table>
  
<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/chemicals?portal=MCF10A

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-chemicals}

portal_name = "MCF10A"

# url for local testing
url0 <- paste0("https://montilab.bu.edu/Xposome-API/chemicals?portal=", portal_name)

# Send GET Request to API
res <- GET(url = url0, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  chemicals <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(chemicals))
  
}

```

### Statistics

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button" id="get-expression-stat" onclick="ToggleOperation()">GET</button></td>
    <td>/gene_expression</td>
    <td>Return  a collection of differential expressed genes that exposed to a known toxin</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/gene_expression?portal=MCF10A&chemical_id=1,2-Dibromo-3-chloropropane&summarize.func=median&landmark=TRUE&do.markers=TRUE&do.scorecutoff=TRUE

<br>
   
<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
  <tr>
    <td>chemical_id</td>
    <td>1,2-Dibromo-3-chloropropane or ...</td>
    <td>Name of the chemicals or CAS ids of a chemical in a specific portal</td>
    <td>string</td>
  </tr>    
  <tr>
    <td>summarize.func</td>
    <td>median (default) or mean/min/max/Q1/Q3</td>
    <td>Name of the summarize functions</td>
    <td>string</td>
  </tr>
  <tr>
    <td>landmark</td>
    <td>TRUE or FALSE (default)</td>
    <td>Whether to include landmark genes</td>
    <td>boolean</td>
  </tr>
  <tr>
    <td>do.nmarkers</td>
    <td>TRUE (default) or FALSE</td>
    <td>Whether to filter by the number of landmark genes (ranging from 1 to 1000)</td>
    <td>boolean</td>
  </tr>  
  <tr>
    <td>do.scorecutoff</td>
    <td>TRUE (default) or FALSE</td>
    <td>Whether to filter by the z-scores (cutoff from -2 to 2)</td>
    <td>boolean</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-gene-expression-stat}

chemical_id = chemicals[4]

# url for local testing
url1 <- paste0("https://montilab.bu.edu/Xposome-API/gene_expression?portal=", portal_name, "&chemical_id=", chemical_id, "&summarize.func=median&landmark=TRUE&do.markers=TRUE&do.scorecutoff=TRUE")

# Send GET Request to API
res <- GET(url = url1, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  gene_expression_stat <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(gene_expression_stat))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/gs_enrichment</td>
    <td>Return  a collection of gene set enrichment that was exposed to a given drug</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/gs_enrichment?portal=MCF10A&chemical_id=1,2-Dibromo-3-chloropropane&geneset=Hallmark&gsva=gsva&summarize.func=median

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
  <tr>
    <td>chemical_id</td>
    <td>1,2-Dibromo-3-chloropropane or ...</td>
    <td>Name of the chemicals or CAS ids of a chemical in a specific portal</td>
    <td>string</td>
  </tr>    
  <tr>
    <td>geneset</td>
    <td>Hallmark (default) or C2 or NURSA</td>
    <td>Collection of the gene set enrichment</td>
    <td>string</td>
  </tr>
  <tr>
    <td>gsva</td>
    <td>gsva (default)</td>
    <td>Method of the gene set enrichment analysis</td>
    <td>string</td>
  </tr>
  <tr>
    <td>summarize.func</td>
    <td>median (default) or mean/min/max/Q1/Q3</td>
    <td>Name of the summarize functions</td>
    <td>string</td>
  </tr>  
</table>

<br>

#### Example in R

```{r get-gene-set-enrichment-stat}

# url for local testing
url2 <- paste0("https://montilab.bu.edu/Xposome-API/gs_enrichment?portal=", portal_name, "&chemical_id=", chemical_id, "&geneset=Hallmark&gsva=gsva&summarize.func=median")

# Send GET Request to API
res <- GET(url = url2, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  gs_enrichment_stat <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(gs_enrichment_stat))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/connectivity</td>
    <td>Return  a collection of human diseases that are linked to exposure of a known chemical</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/connectivity?portal=MCF10A&chemical_id=1,2-Dibromo-3-chloropropane&connectivity_classes=pcl&summarize.func=median

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
  <tr>
    <td>chemical_id</td>
    <td>1,2-Dibromo-3-chloropropane or ...</td>
    <td>Name of the chemicals or CAS ids of a chemical in a specific portal</td>
    <td>string</td>
  </tr>    
  <tr>
    <td>connectivity_classes</td>
    <td>pcl (Perturbagen Classes, default) or pert (Perturbagens)</td>
    <td>Name of the connectivity classes</td>
    <td>string</td>
  </tr>
  <tr>
    <td>summarize.func</td>
    <td>median (default) or mean/min/max/Q1/Q3</td>
    <td>Name of the summarize functions</td>
    <td>string</td>
  </tr>  
</table>

<br>

#### Example in R

```{r get-connecitivity-stat}

# url for local testing
url3 <- paste0("https://montilab.bu.edu/Xposome-API/connectivity?portal=", portal_name, "&chemical_id=", chemical_id, "&connectivity_classes=pcl&summarize.func=median")

# Send GET Request to API
res <- GET(url = url3, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  connectivity_stat <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(connectivity_stat))
  
}

```

### Datasets

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/profile_annotation</td>
    <td>Return the profile annotation of a specific portal</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/profile_annotation?portal=MCF10A

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-profile-annotation}

# url for local testing
url4 <- paste0("https://montilab.bu.edu/Xposome-API/profile_annotation?portal=", portal_name)

# Send GET Request to API
res <- GET(url = url4, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  profile_annotation <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(profile_annotation))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/chemical_annotation</td>
    <td>Return the chemical annotation of of a specific portal</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/chemical_annotation?portal=MCF10A

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-chemical-annotation}

# url for local testing
url5 <- paste0("https://montilab.bu.edu/Xposome-API/chemical_annotation?portal=", portal_name)

# Send GET Request to API
res <- GET(url = url5, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  chemical_annotation <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(chemical_annotation))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/expression_set</td>
    <td>Return the gene expression dataset of a specific portal</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/expression_set?portal=MCF10A

<br>

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-gene-expression}

# url for local testing
url6 <- paste0("https://montilab.bu.edu/Xposome-API/expression_set?portal=", portal_name)

# Send GET Request to API
res <- GET(url = url6, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  expression_set <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(expression_set))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/enrichment_set</td>
    <td>Return the enrichment dataset of a specific portal</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/enrichment_set?portal=MCF10A&geneset=Hallmark&gsva=gsva

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
  <tr>
    <td>geneset</td>
    <td>Hallmark (default) or C2 or NURSA</td>
    <td>Collection of the gene set enrichment</td>
    <td>string</td>
  </tr>
  <tr>
    <td>gsva</td>
    <td>gsva (default)</td>
    <td>Method of gene set enrichment analysis</td>
    <td>string</td>
  </tr>
</table>

<br>

#### Example in R

```{r get-gene-set-entrichment}

# url for local testing
url7 <- paste0("https://montilab.bu.edu/Xposome-API/enrichment_set?portal=", portal_name, "&geneset=Hallmark&gsva=gsva")

# Send GET Request to API
res <- GET(url = url7, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  enrichment_set <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(enrichment_set))
  
}

```

<table class="get-table">
  <tr>
    <td><button type="button" class="btn btn-default action-button">GET</button></td>
    <td>/connectivity_set</td>
    <td>Return the connectivity dataset of a specific portal</td>
  </tr>
</table>

<br>

#### Implementation
https://montilab.bu.edu/Xposome-API/connectivity_set?portal=MCF10A&connectivity_classes=pcl

<table class="api-table">
  <tr>
    <td><strong>Parameter</strong></td>
    <td><strong>Value</strong></td>
    <td><strong>Description</strong></td>
    <td><strong>Data Type</strong></td>
  </tr>
  <tr>
    <td>portal</td>
    <td>MCF10A or HEPG2 or ...</td>
    <td>Name of the portals</td>
    <td>string</td>
  </tr>
  <tr>
    <td>connectivity_classes</td>
    <td>pcl (Perturbagen Classes, default) or pert (Perturbagens)</td>
    <td>Name of the connectivity classes</td>
    <td>string</td>
  </tr>
</table>

#### Example in R

```{r get-connectivity}

# url for local testing
url8 <- paste0("https://montilab.bu.edu/Xposome-API/connectivity_map?portal=", portal_name, "&connectivity_classes=pcl")

# Send GET Request to API
res <- GET(url = url8, encode = 'json')

# Check the status of GET request 
test_request <- tryCatch({
  
  stop_for_status(res)
  
  "pass"
  
}, error = function(e) {
  
  "fail"
  
})

```

#### Output

```{r}

# If GET request is successful, return the results
if(test_request == "pass"){
  
  connectivity_map <- fromJSON(fromJSON(rawToChar(res$content)))
  print(head(connectivity_map))

}

```
